Sources/Tools/SoundDecompress/plOldAudioFileReader.cpp

```diff
@@ -0,0 +1,177 @@
+
+    if (type == kStreamWAV)
+    {
+        bool isWav = (ext.compare_i("wav") == 0);
+        // We want to stream a wav off disk, but this is a compressed file.
+        // Get the uncompressed path. Ignore the requested channel, since it 
+        // will have already been split into two files if that is necessary.
+        if (!isWav)
+        {
+            plFileName cachedPath = IGetCachedPath(path, whichChan);
+            plAudioFileReader *r =  new plFastWAV(cachedPath, plAudioCore::kAll);
+            if (!r->IsValid()) {
+                // So we tried to play a cached file and it didn't exist
+                // Oops... we should cache it now
+                delete r;
+                ICacheFile(path, true, whichChan);
+                r = new plFastWAV(cachedPath, plAudioCore::kAll);
+            }
+            return r;
+        }
+
+        plAudioFileReader *r =  new plFastWAV(path, whichChan);
+        return r;
+    }
+    else if (type == kStreamRAM)
+        return new plBufferedFileReader(path, whichChan);
+    else if (type == kStreamNative)
+        return new plOGGCodec(path, whichChan);
+
+    return nil;
+}
+
+plAudioFileReader* plOldAudioFileReader::CreateWriter(const plFileName& path, plWAVHeader& header)
+{
+#ifdef HS_BUILD_FOR_WIN32
+    plAudioFileReader* writer = new CWaveFile(path.AsString().c_str(), plAudioCore::kAll);
+    writer->OpenForWriting(path.AsString().c_str(), header);
+    return writer;
+#else
+    hsAssert(false, "WAV files can only be generated on Windows");
+    return nullptr;
+#endif
+}
+
+plFileName plOldAudioFileReader::IGetCachedPath(const plFileName& path, plAudioCore::ChannelSelect whichChan)
+{
+    // Get the file's path and add our streaming cache folder to it
+    plFileName cachedPath = plFileName::Join(path.StripFileName(), kCacheDirName);
+
+    // Create the directory first
+    plFileSystem::CreateDir(cachedPath);
+
+    const char *suffix = "";
+    if (whichChan == plAudioCore::kLeft)
+        suffix = "-Left.wav";
+    else if (whichChan == plAudioCore::kRight)
+        suffix = "-Right.wav";
+    else if (whichChan == plAudioCore::kAll)
+        suffix = ".wav";
+
+    // Get the path to the cached version of the file, without the extension
+    return plFileName::Join(cachedPath, path.GetFileNameNoExt() + suffix);
+}
+
+void plOldAudioFileReader::ICacheFile(const plFileName& path, bool noOverwrite, plAudioCore::ChannelSelect whichChan)
+{
+    plFileName cachedPath = IGetCachedPath(path, whichChan);
+    if (!noOverwrite || !plFileInfo(cachedPath).Exists())
+    {
+        plAudioFileReader* reader = plAudioFileReader::CreateReader(path, whichChan, kStreamNative);
+        if (!reader || !reader->IsValid())
+        {
+            delete reader;
+            return;
+        }
+        plAudioFileReader* writer = CreateWriter(cachedPath, reader->GetHeader());
+        if (!writer || !writer->IsValid())
+        {
+            delete reader;
+            delete writer;
```

Comment:
These never escape the current function, and there are several deletes...  They look like a prime candidate for `std::unique_ptr`